sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Cast.java

```diff
@@ -0,0 +1,582 @@
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.Row;
+
+/** Set of utilities for casting rows between schemas. */
+@Experimental(Experimental.Kind.SCHEMAS)
+@AutoValue
+public abstract class Cast<T> extends PTransform<PCollection<T>, PCollection<Row>> {
+
+  public abstract Schema outputSchema();
+
+  public abstract Nullability nullability();
+
+  public abstract Type type();
+
+  public abstract Shape shape();
+
+  /** Builder for {@link Cast}. */
+  @AutoValue.Builder
+  public abstract static class Builder<T> {
+
+    public abstract Builder<T> outputSchema(Schema schema);
+
+    public abstract Builder<T> nullability(Nullability nullability);
+
+    public abstract Builder<T> type(Type type);
+
+    public abstract Builder<T> shape(Shape shape);
+
+    public abstract Cast<T> build();
+  }
+
+  public static <T> Builder<T> builder() {
+    return new AutoValue_Cast.Builder<T>();
+  }
+
+  public static <T> Cast<T> to(Schema outputSchema) {
+    return Cast.<T>builder()
+        .outputSchema(outputSchema)
+        .nullability(Nullability.IGNORE)
+        .type(Type.WIDEN)
+        .shape(Shape.PROJECTION)
+        .build();
+  }
+
+  public List<CompatibilityError> compatibility(Schema inputSchema) {
+    return Inference.compatibility(inputSchema, outputSchema(), nullability(), type(), shape());
+  }
+
+  public void verifyCompatibility(Schema inputSchema) {
+    List<CompatibilityError> errors = compatibility(inputSchema);
+
+    if (!errors.isEmpty()) {
+      String reason =
+          errors
+              .stream()
+              .map(x -> Joiner.on('.').join(x.path()) + ": " + x.message())
+              .collect(Collectors.joining("\n\t"));
+
+      throw new IllegalArgumentException("Cast isn't compatible:\n\t" + reason);
+    }
+  }
+
+  @Override
+  public PCollection<Row> expand(PCollection<T> input) {
+    Schema inputSchema = input.getSchema();
+
+    verifyCompatibility(inputSchema);
+
+    return input
+        .apply(
+            ParDo.of(
+                new DoFn<T, Row>() {
+                  // TODO: This should be the same as resolved so that Beam knows which fields
+                  // are being accessed. Currently Beam only supports wildcard descriptors.
+                  // Once BEAM-4457 is fixed, fix this.
+                  @FieldAccess("filterFields")
+                  final FieldAccessDescriptor fieldAccessDescriptor =
+                      FieldAccessDescriptor.withAllFields();
+
+                  @ProcessElement
+                  public void process(@FieldAccess("filterFields") Row row, OutputReceiver<Row> r) {
+                    r.output(castRow(row, inputSchema, outputSchema()));
+                  }
+                }))
+        .setRowSchema(outputSchema());
+  }
+
+  @VisibleForTesting
+  static Row castRow(Row input, Schema inputSchema, Schema outputSchema) {
```

Comment:
I think that `cast(value, srcType, targetType)` would be a generically useful utility method to make public, then this can be a convenience transform that just maps it over the rows and perhaps has a "dead letter" output for failures (overflow and null cases).

Reply:
@kennknowles I was thinking about it, but I'm not sure it's stable enough. What is the policy regarding version compatibility?

Reply:
All of schemas are still `@Experimental`. I don't think I would commit them to full compatibility until we have the portable semantics fleshed out. I also think the approach to metadata needs some attention; in particular SQL should not use the metadata but trust its compilation process, at which point users probably don't need the metadata at all. Of course, we should save breaking changes until we have worked through all this.